src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala

```diff
@@ -199,44 +199,75 @@ private[internal] trait TypeConstraints {
         val bound: Type = if (up) tparam.info.bounds.hi else tparam.info.bounds.lo
         //Console.println("solveOne0(tv, tp, v, b)="+(tvar, tparam, variance, bound))
         var cyclic = bound contains tparam
+        // This is attempting to reorder the solving of TypeVars such that tvs which are
+        // mentioned in the active bound are solved first.
+        // 1. Is this reordering desirable? Should solving always be left to right instead?
+        // 2. Should this take account of the variance of tparam2 and flip the bounds if
+        //    necessary?
         foreach3(tvars, tparams, variances)((tvar2, tparam2, variance2) => {
-          val ok = (tparam2 != tparam) && (
-            (bound contains tparam2)
-              ||  up && (tparam2.info.bounds.lo =:= tparam.tpeHK)
-              || !up && (tparam2.info.bounds.hi =:= tparam.tpeHK)
-            )
+          val ok = (tparam2 != tparam) && ((bound contains tparam2) || {
+            val bound2 = if (up) tparam2.info.bounds.lo else tparam2.info.bounds.hi
+            bound2 contains tparam
+          })
           if (ok) {
             if (tvar2.constr.inst eq null) cyclic = true
             solveOne(tvar2, tparam2, variance2)
           }
         })
+
         if (!cyclic) {
+          def propagateBounds(tparam2: Symbol): Unit = {
+            def addBound(variance: Variance, bound: Type) = if (bound != NoType) {
+              debuglog(s"$tvar.addBound($variance, $bound.instantiateTypeParams($tparams, $tvars))")
+              if (variance.isInvariant || variance.isCovariant) tvar addHiBound bound
+              if (variance.isInvariant || variance.isContravariant) tvar addLoBound bound
+            }
+
+            def toInst(tp: Type): Type = tp match {
+              case tv: TypeVar =>
+                val inst = tv.inst
+                if (inst == null) NoType else inst
+              case _ => tp
+            }
+
+            // Try harder to propagate information we can discover from this type variable's
+            // occurence in bounds on other type variables. Also avoid some ill-kinding issues.
+            def propagate(variance: Variance, bound: Type, instantiate: => Type): Unit = bound match {
+              // Propagate bounds of the same kind: F[A, B, C[_]] <: G[A, B, C] => F <: G
+              // Undoing eta-expansion to ensure the propagated bound is correct.
+              case GenPolyType(tparams, TypeRef(_, `tparam`, targs))
+                if tparams.corresponds(targs) { (tp, ta) => ta.typeArgs.isEmpty && ta.typeSymbolDirect == tp }
+                => addBound(variance, instantiate)
+              // Propagate type constructor: F[A] <: B => F <: B.typeConstructor
+              // Propagate type arguments: F[+A, -B] <: C => A <: C.typeArgs(0) && B >: C.typeArgs(1)
+              case TypeRef(_, tsym, targs) =>
+                val inst = toInst(instantiate)
```

Comment:
Check for `NoType`?

Reply:
`addBound` checks for `NoType`

Reply:
Just added a check for `NoType` returned from `toInst`.